crates/matrix-sdk-crypto/src/types/mod.rs

```diff
@@ -648,12 +657,66 @@ pub enum ProcessedToDeviceEvent {
     Invalid(Raw<AnyToDeviceEvent>),
 }
 
+// Custom deserialization function for the `Decrypted` variant that
+// support migration of the old format
+fn deserialize_decrypted_variant<'de, D>(
+    deserializer: D,
+) -> Result<(Raw<AnyToDeviceEvent>, EncryptionInfo), D::Error>
+where
+    D: Deserializer<'de>,
+{
+    use serde_json::Value;
+
+    // Deserialize into a generic JSON value to inspect the structure
+    let json = Value::deserialize(deserializer)?;
+
+    // In the old format the Raw<AnyToDeviceEvent> was just flatten out here.
+    // With the new format the event is encapsulated under `raw` and there is a
+    // `encryption_info` field
+    if json.get("raw").is_some() && json.get("encryption_info").is_some() {
+        // If the "raw" field is present, it's a NewFormat
+        #[derive(Deserialize)]
+        struct NewFormat {
+            raw: Raw<AnyToDeviceEvent>,
+            encryption_info: EncryptionInfo,
+        }
```

Comment:
What happens if someone puts a `raw` and `encryption_info` field into the JSON of the event?

It seems to me that the sender of the event, perhaps in collusion with the homeserver, might insert an `encryption_info` on their own.

Come to think of, why does this need to support `Serialize` and `Deserialize`? Because the widgets API will transfer this verbatim to the widget using JSON?

In any case, I think it's safer to not have a migration here since we don't seem to need it.

Reply:
> Come to think of, why does this need to support Serialize and Deserialize?

I'll check but I think it mostly to export then to bindings?
But I agree that those should not be serialized, but we previously said that if we put something as serializable we should provide migration. I wish we had another trait for things that can be persisted